### PR TITLE
ref: Implement the new debug identifiers and kill UUIDs

### PR DESCRIPTION
**Pull request by jan-auer**
**Attention: Breaking Change!**

This migrates from UUIDs and Breakpad-like identifiers to unified debug identifiers, called `ObjectId`. It is a computed *24-byte* identifier with a guaranteed stable but irreversible (i.e. lossy) conversion from Mach-O, ELF, PDB and Breakpad symbols and all future supported formats. To the user, the contents of this identifier are implementation-defined and of no concern. 

## Changes

- [x] Support the new parsing and stringification (`ObjectId`)
- [x] Create a wrapper type for Breakpad IDs (`CodeModuleId`)
- [x] Kill all usages of UUIDs in the public API
- [x] Create Symcache V2 that uses these identifiers
- [x] Update CABI and Python to use strings instead of UUIDs
- [x] Fix all usages, examples and tests

## Details

Internally, it is structured into a 16-byte UUID-like identifier and a 8-byte `appendix` field (which is currently used to hold up to u32 of age data). Parsing and serialization assumes the following:

- The ID is serialized as lowercase UUID with dashes if and only if the appendage field is zero (0).
- The ID is serialized as lowercase UUID with dashes, followed by a dash and a lowercase hex representation of the appendix if and only if it is not zero (0). The appendix is never padded. 

This means the string representation of the ID is between 36 (no appendix) and 53 characters long.

## Examples

**Breakpad:**

With appendix:
- Breakpad ID: `5C80E27984E44E089753ADA08475131Cfeed`
- DIF ID: `5c80e279-84e4-4e08-9753-ada08475131c-feed`

Without appendix:
- Breakpad ID: `5C80E27984E44E089753ADA08475131C0`
- DIF ID: `5c80e279-84e4-4e08-9753-ada08475131c`

**MachO**:

- UUID: `5C80E279-84E4-4E08-9753-ADA08475131C`
- DIF ID: `5c80e279-84e4-4e08-9753-ada08475131c`

**ELF**

- Build ID: `bcfc66cdc7d5147b53b1101194868ef94fe8dddb`
- DIF ID: `cd66fcbc-d5c7-7b14-53b1-101194868ef9`

**PDB**

- File: `crash.pdb`, GUID: `87C94EEA84774851A70AF81BC742CD3A`, Age: `1`
- DIF ID: `87c94eea-8477-4851-a70a-f81bc742cd3a-1`

## Remarks

All new endpoints in Sentry, most importantly the chunk upload endpoint, all internal components, the new symbolserver, symbolic, and new commands in sentry-cli, need to require the new identifier in future. The legacy upload endpoint as well as the corresponding sentry-cli command `upload-dif` continue to use UUIDs. Internally, a loss-less conversion from UUID to identifier is performed. This ensures that sentry and sentry-cli can be updated independently.

The `sentry_projectdsymfile` table needs to be migrated in order to store this new identifier. To be more future-proof, the `uuid` column will be migrated to `VARCHAR(64)`, even though 53 characters will suffice to fit the new identifiers. In the future, we might add more components to this identifier and even further extend its length.

### CODE REVIEW

**Comment by mitsuhiko** [comment]
Since we dump these ids into the symcache format I wonder if we should reserve a second u64 just in case.

**Comment by jan-auer** [comment]
You mean as spacing or what would it be for? Note that the `u64` is already half-empty currently...
Also, still need to update tests in `py/tests` - the rest should be good to go.

**Comment by mitsuhiko** [comment]
@jan-auer just as spacing. I was wondering what happens if we ever need to store a sha256 for a true gnu build note. That would be 32 bytes in total.

**Comment by jan-auer** [comment]
@mitsuhiko Okay, so in that sense we could just create a 32 byte buffer for the IDs and be done with it. Problem is that we're stuck with Breakpad-compatible IDs, so there would be no value in storing the GNU build id.

**Comment by jan-auer** [comment]
@mitsuhiko This is ready now.

File: debuginfo/src/breakpad.rs [hunk 1]
```diff
@@ -5,34 +5,107 @@ use uuid::Uuid;
...
+///
+/// # fn foo() -> Result<()> {
+/// let id = ObjectId::from_breakpad("DFB8E43AF2423D73A453AEB6A777EF75a")?;
+/// assert_eq!("DFB8E43AF2423D73A453AEB6A777EF75a".to_string(), id.breakpad().to_string());
+/// # Ok(())
+/// # }
+///
+/// # fn main() { foo().unwrap() }
+/// ```
+pub trait BreakpadId {
```
Comment (mitsuhiko) [inline debuginfo/src/breakpad.rs]: I want to kill this trait and have it directly on the object id.

File: symcache/src/writer.rs [hunk 2]
```diff
@@ -25,29 +25,14 @@ use utils::shorten_filename;
...
-    buf.write_all(CacheFileHeader::default().as_bytes())?;
-    let header = {
-        let mut writer = SymCacheWriter::new(&mut buf);
-        writer.write_object(obj)?;
-        writer.header
-    };
-    let header_bytes = header.as_bytes();
-    (&mut buf[..header_bytes.len()]).copy_from_slice(header_bytes);
-    Ok(buf)
+    let mut cursor = Cursor::new(Vec::new());
```
Comment (mitsuhiko) [inline symcache/src/writer.rs]: AFAIK we need to revert this to original code because this has a massive performance degradation.

Comment (jan-auer) [reply symcache/src/writer.rs]: Just ran a couple of manual benchmarks, no measurable difference in performance. For some reason, the `Cursor`-based implementation was even 3-4% faster, but I think that is just an error of measurement.

**Pull request merged**